File: postgremq-ts/src/metrics.ts
import { Attributes, Histogram, Counter, UpDownCounter, MeterProvider } from '@opentelemetry/api';
import { performance } from 'perf_hooks';
import {
  ConnectionClosedError,
  LeaseLostError,
  QueueNotFoundError,
  QueueFatalError,
  ValidationError,
  mapDbError,
} from './errors';

function attributes(operation: string, destination: string): Attributes {
  const type =
    operation === 'publish'
      ? 'send'
      : operation === 'consume'
        ? 'receive'
        : ['ack', 'nack', 'release'].includes(operation)
          ? 'settle'
          : operation;
  return {
    'messaging.system': 'postgremq',
    'messaging.operation.name': operation,
    'messaging.operation.type': type,
    ...(destination ? { 'messaging.destination.name': destination } : {}),
  };
}

function errorType(error: unknown): string {
  if (error instanceof LeaseLostError) return 'lease_lost';
  if (error instanceof QueueNotFoundError || error instanceof QueueFatalError)
    return 'queue_not_found';
  if (error instanceof ValidationError) return 'validation';
  if (error instanceof ConnectionClosedError) return 'connection_closed';
  if (error instanceof Error && error.name === 'AbortError') return 'cancelled';
  return 'other';
}

/** Internal implementation of the shared metrics contract in docs/observability.md. */
export class ClientMetrics {
  private readonly renewalLost?: Counter;
  private readonly duration?: Histogram;
  private readonly process?: Histogram;
  private readonly consumed?: Counter;
  private readonly sent?: Counter;
  private readonly active?: UpDownCounter;

  constructor(provider?: MeterProvider) {
    const meter = provider?.getMeter('postgremq', '1');
    const options = {
      unit: 's',
      advice: {
        explicitBucketBoundaries: [
          0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10,
        ],
      },
    };
    this.duration = meter?.createHistogram('messaging.client.operation.duration', options);
    this.process = meter?.createHistogram('messaging.process.duration', options);
    this.sent = meter?.createCounter('messaging.client.sent.messages', { unit: '{message}' });
    this.consumed = meter?.createCounter('messaging.client.consumed.messages', {
      unit: '{message}',
    });
    this.renewalLost = meter?.createCounter('postgremq.client.renewal.lost', { unit: '{message}' });
    this.active = meter?.createUpDownCounter('postgremq.client.handlers.active', {
      unit: '{handler}',
    });
  }

  async operation<T>(
    operation: string,
    destination: string,
    transaction: boolean,
    fn: () => Promise<T>,
  ): Promise<T> {
    if (!this.duration) return fn();
    const start = performance.now();
    const attrs = { ...attributes(operation, destination), 'postgremq.transaction': transaction };
    try {
      return await fn();
    } catch (error) {
      Object.assign(attrs, { 'error.type': errorType(error) });
      throw error;
    } finally {
      this.duration.record((performance.now() - start) / 1000, attrs);
    }
  }

  // Observe the actual driver query attempt, after payload serialization.
  async send<T>(topic: string, transaction: boolean, fn: () => Promise<T>): Promise<T> {
    if (!this.sent) return fn();
    const attrs: Attributes = {
      ...attributes('publish', topic),
      'postgremq.transaction': transaction,
    };
    try {
      return await fn();
    } catch (error) {
      attrs['error.type'] = errorType(mapDbError(error));
      throw error;
    } finally {
      this.sent.add(1, attrs);
    }
  }

  recordConsumed(queue: string, messages: Array<{ delivery_attempts: number }>): void {
    if (!this.consumed) return;
    const redelivered = messages.filter((m) => m.delivery_attempts > 1).length;
    const attrs = attributes('consume', queue);
    if (messages.length > redelivered)
      this.consumed.add(messages.length - redelivered, {
        ...attrs,
        'postgremq.redelivered': false,
      });
    if (redelivered) this.consumed.add(redelivered, { ...attrs, 'postgremq.redelivered': true });
  }

  recordRenewalLost(queue: string): void {
    this.renewalLost?.add(1, {
      'messaging.system': 'postgremq',
      'messaging.destination.name': queue,
    });
  }

  startHandler(queue: string): (errorType: string) => void {
    if (!this.process) return () => {};
    const start = performance.now();
    const attrs = attributes('process', queue);
    this.active?.add(1, attrs);
    return (code) => {
      this.active?.add(-1, attrs);
      this.process?.record((performance.now() - start) / 1000, {
        ...attrs,
        ...(code ? { 'error.type': code } : {}),
      });
    };
  }
}

File: postgremq-go/metrics.go
package postgremq_go

import (
	"context"
	"errors"
	"time"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

// WithMeterProvider enables client metrics. The application owns the provider,
// readers, exporters and shutdown. Without this option metrics are disabled,
// even if a global provider is configured. Pass otel.GetMeterProvider() explicitly
// to use the global provider.
func WithMeterProvider(provider metric.MeterProvider) ConnectionOption {
	return func(c *Connection) { c.meterProvider = provider }
}

type clientMetrics struct {
	operation   metric.Float64Histogram
	process     metric.Float64Histogram
	consumed    metric.Int64Counter
	sent        metric.Int64Counter
	active      metric.Int64UpDownCounter
	renewalLost metric.Int64Counter
}

func newClientMetrics(provider metric.MeterProvider) (*clientMetrics, error) {
	if provider == nil {
		return nil, nil
	}
	meter := provider.Meter("postgremq", metric.WithInstrumentationVersion("1"))
	m := &clientMetrics{}
	var err error
	boundaries := metric.WithExplicitBucketBoundaries(.005, .01, .025, .05, .075, .1, .25, .5, .75, 1, 2.5, 5, 7.5, 10)
	m.operation, err = meter.Float64Histogram("messaging.client.operation.duration", metric.WithUnit("s"), boundaries)
	if err != nil {
		return nil, err
	}
	m.process, err = meter.Float64Histogram("messaging.process.duration", metric.WithUnit("s"), boundaries)
	if err != nil {
		return nil, err
	}
	m.sent, err = meter.Int64Counter("messaging.client.sent.messages", metric.WithUnit("{message}"))
	if err != nil {
		return nil, err
	}
	m.consumed, err = meter.Int64Counter("messaging.client.consumed.messages", metric.WithUnit("{message}"))
	if err != nil {
		return nil, err
	}
	m.active, err = meter.Int64UpDownCounter("postgremq.client.handlers.active", metric.WithUnit("{handler}"))
	if err != nil {
		return nil, err
	}
	m.renewalLost, err = meter.Int64Counter("postgremq.client.renewal.lost", metric.WithUnit("{message}"))
	if err != nil {
		return nil, err
	}
	return m, nil
}

func metricAttributes(operation, destination string) []attribute.KeyValue {
	kind := operation
	switch operation {
	case "publish":
		kind = "send"
	case "consume":
		kind = "receive"
	case "ack", "nack", "release":
		kind = "settle"
	}
	attrs := []attribute.KeyValue{attribute.String("messaging.system", "postgremq"), attribute.String("messaging.operation.name", operation), attribute.String("messaging.operation.type", kind)}
	if destination != "" {
		attrs = append(attrs, attribute.String("messaging.destination.name", destination))
	}
	return attrs
}

func metricError(err error) string {
	switch {
	case err == nil:
		return ""
	case errors.Is(err, ErrLeaseLost):
		return "lease_lost"
	case errors.Is(err, ErrQueueNotFound), errors.Is(err, ErrQueueGone):
		return "queue_not_found"
	case errors.Is(err, ErrValidation):
		return "validation"
	case errors.Is(err, ErrConnectionClosed):
		return "connection_closed"
	case errors.Is(err, context.Canceled):
		return "cancelled"
	case errors.Is(err, context.DeadlineExceeded):
		return "deadline_exceeded"
	default:
		return "other"
	}
}

// Observe logical operations once, including pool wait and internal retries.
// The transaction attribute describes caller ownership, never commit outcome.
func (m *clientMetrics) startOperation(ctx context.Context, operation, destination string, transaction bool) func(error) {
	if m == nil {
		return func(error) {}
	}
	start := time.Now()
	attrs := append(metricAttributes(operation, destination), attribute.Bool("postgremq.transaction", transaction))
	return func(err error) {
		if code := metricError(err); code != "" {
			attrs = append(attrs, attribute.String("error.type", code))
		}
		m.operation.Record(ctx, time.Since(start).Seconds(), metric.WithAttributes(attrs...))
	}
}

func (m *clientMetrics) recordConsumed(ctx context.Context, queue string, messages []*Message, err error) {
	if m == nil {
		return
	}
	var redelivered int64
	for _, msg := range messages {
		if msg.DeliveryAttempt > 1 {
			redelivered++
		}
	}
	attrs := metricAttributes("consume", queue)
	if code := metricError(err); code != "" {
		attrs = append(attrs, attribute.String("error.type", code))
	}
	for _, entry := range []struct {
		count       int64
		redelivered bool
	}{{int64(len(messages)) - redelivered, false}, {redelivered, true}} {
		if entry.count > 0 {
			m.consumed.Add(ctx, entry.count, metric.WithAttributes(append(attrs, attribute.Bool("postgremq.redelivered", entry.redelivered))...))
		}
	}
}

func (m *clientMetrics) startHandler(ctx context.Context, queue string) func(string) {
	if m == nil {
		return func(string) {}
	}
	attrs := metricAttributes("process", queue)
	start := time.Now()
	m.active.Add(ctx, 1, metric.WithAttributes(attrs...))
	return func(code string) {
		m.active.Add(ctx, -1, metric.WithAttributes(attrs...))
		if code != "" {
			attrs = append(attrs, attribute.String("error.type", code))
		}
		m.process.Record(ctx, time.Since(start).Seconds(), metric.WithAttributes(attrs...))
	}
}

// Recorded only when the renewal actor retires a still-tracked delivery.
func (m *clientMetrics) recordRenewalLost(queue string) {
	if m == nil {
		return
	}
	m.renewalLost.Add(context.Background(), 1, metric.WithAttributes(attribute.String("messaging.system", "postgremq"), attribute.String("messaging.destination.name", queue)))
}

// Count a message when a publish SQL attempt finishes, including failed attempts.
// Retries are separate attempts; transaction success does not imply commit.
func (m *clientMetrics) recordSent(ctx context.Context, topic string, transaction bool, err error) {
	if m == nil {
		return
	}
	attrs := append(metricAttributes("publish", topic), attribute.Bool("postgremq.transaction", transaction))
	if code := metricError(mapPgError(err)); code != "" {
		attrs = append(attrs, attribute.String("error.type", code))
	}
	m.sent.Add(ctx, 1, metric.WithAttributes(attrs...))
}
